Implement asynchronous eventSender mouse functions and begin adopting them

charliewolfe opened the pull request:
#### 6310806e67493a9e1167b9d64e312a0553b568ec
<pre>
Implement asynchronous eventSender mouse functions and begin adopting them
<a href="https://bugs.webkit.org/show_bug.cgi?id=277172">https://bugs.webkit.org/show_bug.cgi?id=277172</a>
<a href="https://rdar.apple.com/132597860">rdar://132597860</a>

Reviewed by Alex Christensen.

When running layout tests with `—site-isolation`, many tests that use `eventSender` fail because iframe
contents may shift between the time the `eventSender` function is invoked and when the iframe process
receives the event. This happens because, with site isolation, input events need to be sent to the
mainframe first before they can be sent to a child frame, and the additional asynchronous IPC message
means that web content could change before the iframe receives the event.

To fix this we can add asynchronous event sending functions that will await until all web processes have
handled the event.

* LayoutTests/fast/events/autoscroll-in-iframe-body.html:
* LayoutTests/fast/events/autoscroll-in-iframe.html:
* LayoutTests/fast/events/pointer-events-2.html:
* LayoutTests/fast/events/pointer-events.html:
* LayoutTests/fast/events/related-target.html:
* LayoutTests/fast/forms/25153.html:
* LayoutTests/fast/forms/range/slider-mouse-events.html:
* LayoutTests/fast/forms/search-rtl.html:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(+[EventSendingController isSelectorExcludedFromWebScript:]):
(+[EventSendingController webScriptNameForSelector:]):
(-[EventSendingController asyncMouseDown:withModifiers:]):
(-[EventSendingController asyncMouseUp:withModifiers:]):
(-[EventSendingController asyncMouseMoveToX:Y:]):
* Tools/WebKitTestRunner/EventSenderProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/EventSendingController.idl:
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::asyncMouseDown):
(WTR::EventSendingController::asyncMouseUp):
(WTR::EventSendingController::asyncMouseMoveTo):
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.h:
* Tools/WebKitTestRunner/InjectedBundle/ios/EventSenderProxyIOS.mm:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
* Tools/WebKitTestRunner/gtk/EventSenderProxyGtk.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/libwpe/EventSenderProxyLibWPE.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::EventSenderProxy::mouseDown):
(WTR::EventSenderProxy::mouseUp):
(WTR::EventSenderProxy::mouseMoveTo):
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/win/EventSenderProxyWin.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):

Canonical link: <a href="https://commits.webkit.org/281547@main">https://commits.webkit.org/281547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77d9aacafb4459e5314c0882b173c24f02e18d8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64200 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10812 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62410 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47304 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11045 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48810 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7523 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62311 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52218 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29653 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9446 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9729 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65932 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4214 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52194 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56332 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3508 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9040 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/WebKitTestRunner/EventSenderProxy.h:
```diff
@@ -147,6 +150,7 @@ class EventSenderProxy {
     unsigned m_mouseButtonsCurrentlyDown { 0 };
 #if PLATFORM(COCOA)
     int m_eventNumber { 0 };
+    NSView *m_targetView { nullptr };
```
WeakObjCPtr or RetainPtr

achristensen07 commented on Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
```diff
@@ -950,4 +950,15 @@ static CGMomentumScrollPhase cgMomentumPhaseFromPhase(EventSenderProxy::WheelEve
+        [checked_objc_cast<WKWebView>(targetView) _doAfterProcessingAllPendingMouseEvents:^{
+            doneProcessingMouseEvents = true;
+        }];
+        m_testController->runUntil(doneProcessingMouseEvents, 100_ms);
```
I imagine we'll want to wait forever, right?

charliewolfe replied:
There are rare situations that may cause this callback to not be called, like if the WP exits before the UI process receives a reply that the mouse event was handled.

charliewolfe committed:
Implement asynchronous eventSender mouse functions and begin adopting them
https://bugs.webkit.org/show_bug.cgi?id=277172
rdar://132597860

Reviewed by Alex Christensen.

When running layout tests with `—site-isolation`, many tests that use `eventSender` fail because iframe
contents may shift between the time the `eventSender` function is invoked and when the iframe process
receives the event. This happens because, with site isolation, input events need to be sent to the
mainframe first before they can be sent to a child frame, and the additional asynchronous IPC message
means that web content could change before the iframe receives the event.

To fix this we can add asynchronous event sending functions that will await until all web processes have
handled the event.

* LayoutTests/fast/events/autoscroll-in-iframe-body.html:
* LayoutTests/fast/events/autoscroll-in-iframe.html:
* LayoutTests/fast/events/pointer-events-2.html:
* LayoutTests/fast/events/pointer-events.html:
* LayoutTests/fast/events/related-target.html:
* LayoutTests/fast/forms/25153.html:
* LayoutTests/fast/forms/range/slider-mouse-events.html:
* LayoutTests/fast/forms/search-rtl.html:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(+[EventSendingController isSelectorExcludedFromWebScript:]):
(+[EventSendingController webScriptNameForSelector:]):
(-[EventSendingController asyncMouseDown:withModifiers:]):
(-[EventSendingController asyncMouseUp:withModifiers:]):
(-[EventSendingController asyncMouseMoveToX:Y:]):
* Tools/WebKitTestRunner/EventSenderProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/EventSendingController.idl:
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::asyncMouseDown):
(WTR::EventSendingController::asyncMouseUp):
(WTR::EventSendingController::asyncMouseMoveTo):
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.h:
* Tools/WebKitTestRunner/InjectedBundle/ios/EventSenderProxyIOS.mm:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
* Tools/WebKitTestRunner/gtk/EventSenderProxyGtk.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/libwpe/EventSenderProxyLibWPE.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::EventSenderProxy::mouseDown):
(WTR::EventSenderProxy::mouseUp):
(WTR::EventSenderProxy::mouseMoveTo):
(WTR::EventSenderProxy::waitForPendingMouseEvents):
* Tools/WebKitTestRunner/win/EventSenderProxyWin.cpp:
(WTR::EventSenderProxy::waitForPendingMouseEvents):

Canonical link: https://commits.webkit.org/281547@main
